## Task 2
### In this task you will work with data collected from our wheeled robot , check How_data_collected_for_Task_2.ipynb for more info.
### You have 7 txt files (0.txt to 6.txt) each represent a step
### build a map of the robot at each step independently using the code provided in Task 1 ( taken from http://ais.informatik.uni-freiburg.de/teaching/ss20/robotics/)

In [83]:
import numpy as np
import matplotlib.pyplot as plt
import math
import cv2
import csv
pi = math.pi
#arr[]=np.array()

In [84]:
for i in range(7):### loop opens each file and prints number of measurments
    
    with open(str(i)+'.txt') as f:
        line_1 = f.readline()
        line_as_list=line_1.split('\t')
        arr = np.array(line_as_list)
        print(len(arr))
    f.close()
##### print  

100
73
107
66
122
99
56


In [85]:

## oppen one file and review data in array arr
with open(str(i)+'.txt') as f:
    line_1 = f.readline()
    line_as_list=line_1.split('\t')
    arr = np.array(line_as_list)
    print(len(arr))
    print(arr[0])# 
f.close()


56
(15, 280.296875, 1203.0)


###  What does (15, 280.296875, 1203.0) mean (check https://rplidar.readthedocs.io/en/latest/ for more info)
#### First value (mostly 15 ) is  quality : int which is  Reflected laser pulse strength

#### Second value is  angle : float which is the measurment heading angle in degree unit [0, 360)

#### Third value is  distance : float Measured object distance related to the sensor’s rotation center. In millimeter unit. Set to 0 when measurment is invalid.

In [1]:
## Here comes your code 
## generate occupancy gird map from each step

for i in range(7):
    with open(str(i)+'.txt') as f:
        line_1 = f.readline()
        readings = []
        img = np.zeros((3000,3000),dtype=np.uint8)
        img.fill(128)
        line_as_list=line_1.split('\t')
        arr = np.array(line_as_list)
        for element in arr:
            element = element.replace('(','')
            element = element.replace(')','')
            element = element.split(',')
            readings.append([float(element[1]), float(element[2])])
        for reading in readings:
            x1=1500.0
            y1= 1500.0
            angle = math.radians(reading[0])
            x2 = int (x1 + math.cos(angle)*reading[1])
            y2 = int (y1 + math.sin(-1*angle)*reading[1])
            x1 = int(x1)
            y1= int(y1)
            cv2.line(img,(x1,y1),(x2,y2),(255,255,255),thickness=4)
            cv2.circle(img,(x2,y2),4,(0,0,0),thickness=-1)
            cv2.imwrite(str(i)+'.png', img)
f.close()
